dhall/src/Dhall/Package.hs

```diff
@@ -44,72 +47,152 @@ writePackage characterSet outputFn inputs = do
-            when (dir' /= outputDir') $
+            absoluteDir <- addTrailingPathSeparator <$> makeAbsolute (normalise dir)
+            let relativeDir = makeRelative outputDir' absoluteDir
+            unless (isRelative relativeDir) $
                 throwIO $ AmbiguousOutputDirectory outputDir dir
+            return relativeDir
 
     resultMap <- go Map.empty checkOutputDir (path:paths)
-    return (outputDir </> outputFn', RecordLit resultMap)
+    return (outputDir </> outputFn', RecordLit $ Map.sort resultMap)
     where
-        go :: Map Text (RecordField s Import) -> (FilePath -> IO ()) -> [FilePath] -> IO (Map Text (RecordField s Import))
+        go :: Map Text (RecordField s Import) -> (FilePath -> IO FilePath) -> [FilePath] -> IO (Map Text (RecordField s Import))
         go !acc _checkOutputDir [] = return acc
         go !acc checkOutputDir (p:ps) = do
             isDirectory <- doesDirectoryExist p
             isFile <- doesFileExist p
             if | isDirectory -> do
-                    checkOutputDir p
+                    void $ checkOutputDir p
                     entries <- listDirectory p
                     let entries' = filter (\entry -> takeExtension entry == ".dhall") entries
                     go acc checkOutputDir (map (p </>) entries' <> ps)
                | isFile -> do
-                    checkOutputDir $ takeDirectory p
+                    dir <- checkOutputDir $ takeDirectory p
+
+                    let p' = normalise $ dir </> takeFileName p
+
+                    let resultMap = if p' == outputFn'
+                            then Map.empty
+                            else filepathToMap outputFn' p'
+
+                    acc' <- mergeMaps acc resultMap
+                    go acc' checkOutputDir ps
+                | otherwise -> throwIO $ InvalidPath p
 
-                    let key = Text.pack $ dropExtension $ takeFileName p
+        outputFn' = fromMaybe "package.dhall" outputFn
 
+-- | Construct a nested 'Map' from a 'FilePath'.
+-- For example, the filepath @some/file/path.dhall@ will result in something
+-- similar to the following:
+--
+-- fromList
+--   [ ("some", fromList
+--     [ ("file", fromList
+--       [ ("path", ./some/file/path.dhall)
+--       ])
+--     ])
+--   ])
+--
+-- ... where ./some/file/path.dhall is a Dhall import. If the last component
+-- equals the value passed in the @outputFn@ argument we produce a slightly
+-- different result. Consider for example the Dhall Prelude: We have some
+-- sub-packages there like @List/package.dhall@. If we want to construct the
+-- top-level @package.dhall@ we want an entry like
+--
+-- > List = ./List/package.dhall
+--
+-- in there and not:
+--
+-- > List = { package = ./List/package.dhall }
+--
+filepathToMap :: FilePath -> FilePath -> Map Text (RecordField s Import)
+filepathToMap outputFn = go [] . splitDirectories
+    where
+        go _acc [] = error "filepathToMap: Path is empty"
```

Comment:
```suggestion
        go acc [] = go acc ["."]
```

Even though we currently don't hit this case in practice, it's probably worth handling it anyway in case the rest of the code changes.